### PR TITLE
Functional assay full

### LINKED ISSUE #1000: functional assay final tweaks
This issue serves as a master list of issues to resolve before v1 release of functional assays (related to #941 and #949).

I'll keep http://brcaexchange-dev.gi.ucsc.edu:9003/ up to date periodically as we resolve outstanding tasks, it currently has a dataset generated 2/14/19 including Findlay Ring Function Scores.

@falquaddoomi I combined the functionalAssayBackend and functionalAssayUI branches into https://github.com/BRCAChallenge/brca-exchange/tree/functionalAssayFull so we could review on dev -- do you see any major issues with working on the functionalAssayFull branch to wrap things up?

Please mention any final tweaks required before releasing v1 of the functional assays or feel free to add them to the list below:

- [x] Review data
- [x] Update source filters
- [x] ~~Include column selectors for findlay fields?~~ (see #1001)
- [ ] Update release notes
- [x] Update docs
- [x] Variants with findlay as the only source don't show up in search results (must have other sources as well)

### PR DESCRIPTION
**Pull request by zfisch**
Note: This branch was created by merging the functionalAssayBackend branch together with the functionalAssayUI branch. This PR closes #1000, #941, and #949. 

I'm opening this early so please feel free to make tweaks and/or suggestions as you see fit -- I hope for this to be an opportunity to identify issues and clean things up before merge.

Documentation updates will be made in a separate PR by @amycoffin.

The 3 core parts of this PR are as follows:

1. Download and parse Findlay BRCA1 Ring Function Scores as part of the automated pipeline.
2. Merge Findlay variants with other variants in the automated pipeline.
3. Store and display Findlay results on brcaexchange.

I plan to keep the latest changes up to date on http://brcaexchange-dev.gi.ucsc.edu:9003 until merge.

### CODE REVIEW

File: pipeline/luigi/CompileVCFFiles.py [hunk 1]
```diff
@@ -1245,6 +1246,148 @@ def run(self):
 
         copy(self.output().path, PipelineParams().output_dir)
 
+
```
Comment (zfisch) [inline pipeline/luigi/CompileVCFFiles.py]: Added this to make sure all sources follow the same pattern -- i.e. they all get copied into the output directory.

File: pipeline/functional_assays/parse_functional_assay_data.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    NUM_HEADER_LINES = 2
+    with open(f_in, 'r') as f_in:
+        with open(f_tmp.name, 'w') as f_tmp_opened:
+            while NUM_HEADER_LINES > 0:
+                f_in.next()
+                NUM_HEADER_LINES -= 1
+            for line in f_in:
+                f_tmp_opened.write(line)
+
+    f_in_data_frame = pd.read_csv(f_tmp.name, sep="\t")
```
Comment (izcram) [inline pipeline/functional_assays/parse_functional_assay_data.py]: maybe the `skiprows` parameter in `pd.read_csv` would avoid the temporary file.

File: pipeline/functional_assays/parse_functional_assay_data.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        'hgvs_nucleotide',
+        'log_rna_depletion',
+        'functional_enrichment_score']
+
+    new_data_frame = pd.DataFrame(columns=new_columns)
+
+    new_data_frame = new_data_frame.reset_index(drop=True)
+    f_in_data_frame = f_in_data_frame.reset_index(drop=True)
+
+    new_data_frame.chr = f_in_data_frame['chromosome']
```
Comment (izcram) [inline pipeline/functional_assays/parse_functional_assay_data.py]: you are probably looking for pandas `rename` function. Sth along the lines of `f_in_data_frame.rename(columns={'chromosome' : 'chr', 'position (hg19)' : 'pos_hg19' etc })`

File: pipeline/functional_assays/functional_assays_to_vcf.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        for field in headerline:
+            field_index = fieldIdxDict[field]
+            field_value = parsedLine[field_index]
+            field_value = normalize(field, field_value)
+            INFO_field.append('{0}={1}'.format(field, field_value))
+
+        # extract hgvs cDNA term for variant and cleanup formatting
+        hgvsName = parsedLine[fieldIdxDict['hgvs_nucleotide']]
+        if hgvsName == '-':
+            print(parsedLine)
```
Comment (izcram) [inline pipeline/functional_assays/functional_assays_to_vcf.py]: Is there a `file=` argument missing? If not, maybe using `logging` to clearly separate diagnostics from output.

File: pipeline/functional_assays/parse_functional_assay_data.py (hunk 3 above)
Comment (zfisch) [reply pipeline/functional_assays/parse_functional_assay_data.py]: 👍

File: pipeline/functional_assays/parse_functional_assay_data.py (hunk 2 above)
Comment (zfisch) [reply pipeline/functional_assays/parse_functional_assay_data.py]: indeed, thanks for the tip. done!

File: pipeline/functional_assays/functional_assays_to_vcf.py (hunk 4 above)
Comment (zfisch) [reply pipeline/functional_assays/functional_assays_to_vcf.py]: makes sense, added explicit logging. FWIW this file is a tweaked copy of lovd2vcf.

**Comment by falquaddoomi** [comment]
I wasn't sure if this was intended, but the front-end code was still using a JSON file that I derived from the Lea Starita FA results spreasheet that maps the transcript ID + cDNA string for variants to FA scores. I removed that file since it's a little large (148KB), but since we still need the scores to generate the histogram (for now) I replaced it with a smaller (48KB) file of just the score values. The front-end code now gets the score from the variant payload, specifically the `Functional_Enrichment_Score_Findlay_BRCA1_Ring_Function_Scores` field. I've spot-checked a few values and they line up with what's in the spreadsheet.

Regarding the resource file, ideally we'll eventually either be able to copy the scores file into a frontend-accessible directory, serve it through the API, or possibly generate the histogram SVG in the pipeline or during data deployment.

**Pull request merged**